habit-backend/users/views.py

```diff
@@ -0,0 +1,89 @@
+import re
+
+from django.contrib.auth import authenticate, login
+from rest_framework import permissions, generics, status
+from rest_framework.response import Response
+from rest_framework_jwt.settings import api_settings
+from .userhelper import ValidateEmail
+
+from users.models import User
+from users.serializers import UserSerializer , LoginSerializer
+
+jwt_payload_handler = api_settings.JWT_PAYLOAD_HANDLER
+jwt_encode_handler = api_settings.JWT_ENCODE_HANDLER
+
+
+class SignUpUserView(generics.GenericAPIView):
+    permission_classes = (permissions.AllowAny,)
+
+    def post(self, request, *args, **kwargs):
+
+        email = request.data.get('email')
+        if email is None:
+            return Response({'email': ['This field is required.']},
+                            status=status.HTTP_400_BAD_REQUEST)
+
+        email = email.lower()
+        email_valid_status = ValidateEmail(email)
+        if email_valid_status == "Invalid":
+            return Response({'email': ['This field is invalid.']},
+                            status=status.HTTP_400_BAD_REQUEST)
+
+        password = request.data.get('password')
+        if password is None:
+            return Response({'password': ['This field is required.']},
+                            status=status.HTTP_400_BAD_REQUEST)
+
+
+        first_name = request.data.get('first_name')
+        last_name = request.data.get('last_name')
+
+        if first_name is None:
+            return Response({'first_name': ['This field is required.']},
+                            status=status.HTTP_400_BAD_REQUEST)
+
+        if last_name is None:
+            return Response({'last_name': ['This field is required.']},
+                            status=status.HTTP_400_BAD_REQUEST)
+
+        gender = request.data.get('gender')
+        if gender is None:
+            return Response({'gender': ['This field is required.']},
+                            status=status.HTTP_400_BAD_REQUEST)
+
+        if gender != 'male' and gender != 'female' and gender != 'other':
+            return Response({'gender': ['This field is invalid.']},
+                            status=status.HTTP_400_BAD_REQUEST)
+
+        try:
+            existing_user = User.objects.get(email=email)
```

Comment:
Should we be checking this before data validations?
Is there any advantage checking it later?

Reply:
We could check if user exists or nor after email validation check and rest of the validations we could move in except ? What do you say about that.?

Reply:
Yeah,
Because if the email exists and his first_name contains a number supposedly, or is greater than 50.

We should give him an error about email first, and then the first_name